### DISCUSSION TITLE
Alternative to in state guards

### DISCUSSION THREAD

**Discussion by bluewave41** [statelyai/xstate#4029] [Q&A]
So lets say I have a basic form with a username and password field kind of like the machine shown here

```
states: {
    username: {
        states: {
            empty: {},
            valid: {}
        },
        on: {
            UPDATE_USERNAME: { actions: 'updateUsername' }
        }
    },
    password: {
        states: {
            empty: {},
            valid: {}
        },
        on: {
            UPDATE_PASSWORD: { actions: 'updatePassword' }
        }
    }
    on: {
        SUBMIT: [
            { cond: 'isFormValid', actions: 'next' },
            { target: '.invalid'}
        ]
    }
}
```

What is a good way to do the isFormValid guard to make sure nothing is in the empty state? I've been using a guard kind of like

```
         ['username.valid',
          'password.valid',
        ].every((match) => state.matches(match)),
```

 that checks if the username.valid and password.valid states match but it's been pointed out to me that it's very similar to [in state guards](https://xstate.js.org/docs/guides/guards.html#multiple-guards) and discouraged

Final states don't really work because it's not final. A user always needs to be able to edit these so there's no "end state"

Is there a better alternative to this?

**Comment by chaitanyaj2222** [comment]
Hey @bluewave41 . If you are still looking for a solution to this one, We have designed a whole big form using state machines in my workplace & I could say that this is the better approach for validating the form. Since form validation depends on the validation of every required field in the form, I would say that the logic that you have written in `isFormValid` is the appropriate one.

>that checks if the username.valid and password.valid states match but it's been pointed out to me that it's very similar to [in state guards](https://xstate.js.org/docs/guides/guards.html#multiple-guards) and discouraged

I would like to know why is using `in` state guards discouraged? I couldn't find a reason in the docs.

**Reply by davidkpiano** [reply]
In-state guards are meant for states within parallel states; they're discouraged elsewhere.
